### PR TITLE
Improving stability.

### PR DESCRIPTION
**Pull request by mzhurovich**

### CODE REVIEW

File: Karl/claire.h [hunk 1]
```diff
@@ -280,7 +296,20 @@ class GenericClaire final : private DummyClaireNotifiable {
   }
 
   void StartKeepaliveThread() {
-    keepalive_thread_ = std::thread([this]() { KeepaliveThread(); });
+    std::lock_guard<std::mutex> lock(keepalive_thread_running_status_mutex_);
+    if (destructing_) {
+      return;
+    }
+    if (!keepalive_thread_running_) {
+      keepalive_thread_ = std::thread([this]() { keepalive_thread_running_ = true; KeepaliveThread(); });
```
Comment (mzhurovich) [inline Karl/claire.h]: Indent?

File: Karl/test_service/http_subscriber.h [hunk 2]
```diff
@@ -38,7 +38,14 @@ class HTTPStreamSubscriber {
         callback_(callback),
         destructing_(false),
         index_(0u),
-        thread_([this]() { Thread(); }) {}
+        has_terminate_id_(false),
+        thread_([this]() { Thread(); }) {
+    while (!has_terminate_id_) {
+      // Note: This will hang forever is the local Karl is not responsing.
```
Comment (mzhurovich) [inline Karl/test_service/http_subscriber.h]: s/is the/if the/

File: Karl/claire.h [hunk 3]
```diff
@@ -121,17 +122,29 @@ class GenericClaire final : private DummyClaireNotifiable {
...
-      } catch (net::NetworkException&) {
+    std::unique_lock<std::mutex> lock(keepalive_thread_running_status_mutex_);
+
+    // Signal the keepalive thread to shut down if running, and prevent the one one from starting.
+    destructing_ = true;
+
+    if (keepalive_thread_running_) {
+      if (keepalive_thread_.joinable()) {
+        {
+          std::lock_guard<std::mutex> lock(keepalive_mutex_);
```
Comment (mzhurovich) [inline Karl/claire.h]: s/lock/keepalive_mutex_lock/ pls

**Comment by mzhurovich** [comment]
@dkorolev 👍👍👍

**Comment by dkorolev** [comment]
Overnight results: 250K x 2 iterations of `Karl.{EndToEndTest,KarlNotifiesUserObject}` on Hetz, and counting.

**Pull request merged**